### DISCUSSION TITLE
Repeated time values cause an error when using EVID 4

### DISCUSSION THREAD

**Discussion by slove1200** [metrumresearchgroup/mrgsolve#1005] [Q&A] [answered]
Dear mrgsolve community, 

I ran into a problem regarding the design of time column with the same ID. In my dataset, I have IDs with different week sampling times (Sampling WEEK), and each sampling time I have 2 samples(replicates). For each sample, I'd like to give an AMT = 1 in CMT = 1 at TIME = 0 and calculate some values that I'm interested in during TIME = 0-42. Then I want to reset the compartment and dosing again when there is a new sample. My dataset looks like the table below:

![image](https://user-images.githubusercontent.com/110170808/181497580-2bd9e6ff-53dd-43d4-9a6f-07e52a589ac8.png)

However, an `Error: the data set is not sorted by time` occurs when I run the simulation. Maybe I am a bit confused about how EVID = 4 behaves in mrgsolve, but from what I understand, this would reset the compartment system and give dose again, so I supposed time value can be repeated when I have the same ID? I am wondering whether there's any way of manipulating the dataset or model code to achieve my goal of simulation. 

Thank you in advance for the help!

**Comment by FelicienLL** [comment]
Hello, 
Yes, it does reset the compartment and adds the dose, but time must still be increasing, unfortunately. https://github.com/metrumresearchgroup/mrgsolve/issues/713
My workaround so far is either to generate a new TIME variable (e.g. 0, 1, 2,...42, 1000, 1001, ..., 1042, 2000, ...) or a new ID (1001, 2001, 3001, 4001...).

**Comment by kylebaron** [comment, marked as answer]
Hi @slove1200  - 

Thanks for the question. And thanks to @FelicienLL for help with the answer. 

Right now, time must be increasing. mrgsolve has been this way since the start and it would take some non-trivial changes to get this working with expected behavior for _all_ the ways you can call mrgsim.  

I'm putting some example code below so you can implement a workaround (it's about what I do now when I run into this). 

I'll start working on a function that will do this data modification automatically to (hopefully) avoid most of the inconvenience. And will keep thinking about how we can change the code to have this (eventually) be a non-issue ... but that'll prob take some time. 

Please LMK if this solves the problem for you.

Kyle



``` r
library(mrgsolve)
#> 
#> Attaching package: 'mrgsolve'
#> The following object is masked from 'package:stats':
#> 
#>     filter
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
```

# Model

``` r
mod <- house()
```

# Data set up

You don’t need to do this … just for demo

``` r
e1 <- evd(amt = 100, evid = 1, ID = 1) %>% 
  expand_observations(seq(0,24,8))
e2 <- evd(amt = 200, evid = 4, ID = 1) %>% 
  expand_observations(seq(0,24,6))

data <- rbind(e1, e2)
```

# Data

``` r
as.data.frame(data)
#>    ID TIME AMT CMT EVID
#> 1   1    0   0   0    0
#> 2   1    0 100   1    1
#> 3   1    8   0   0    0
#> 4   1   16   0   0    0
#> 5   1   24   0   0    0
#> 6   1    0   0   0    0
#> 7   1    0 200   1    4
#> 8   1    6   0   0    0
#> 9   1   12   0   0    0
#> 10  1   18   0   0    0
#> 11  1   24   0   0    0
```

# Modify time

``` r
data <- 
  data %>%
  mutate(OTIME = TIME) %>% 
  group_by(ID) %>% 
  mutate(TIME = cumsum(TIME)) %>% 
  ungroup()
```

# Simulate

``` r
out <- mrgsim(mod, data, carry_out = "TAD = OTIME, EVID")
```

# Check

``` r
as.data.frame(out)
#>    ID TIME EVID TAD          GUT       CENT     RESP        DV        CP
#> 1   1    0    0   0 0.000000e+00   0.000000 50.00000 0.0000000 0.0000000
#> 2   1    0    1   0 1.000000e+02   0.000000 50.00000 0.0000000 0.0000000
#> 3   1    8    0   8 6.772876e-03  69.939372 36.80390 3.4969686 3.4969686
#> 4   1   24    0  16 3.333687e-11  31.428961 43.06040 1.5714481 1.5714481
#> 5   1   48    0  24 1.359515e-23   9.466221 47.68521 0.4733111 0.4733111
#> 6   1   48    0   0 1.359515e-23   9.466221 47.68521 0.4733111 0.4733111
#> 7   1   48    4   0 2.000000e+02   0.000000 50.00000 0.0000000 0.0000000
#> 8   1   54    0   6 1.493172e-01 154.449732 27.92163 7.7224866 7.7224866
#> 9   1   66    0  12 8.361770e-08  84.849320 34.84225 4.2424660 4.2424660
#> 10  1   84    0  18 3.717616e-17  34.497160 42.48481 1.7248580 1.7248580
#> 11  1  108    0  24 1.516601e-29  10.390345 47.47067 0.5195172 0.5195172
```

<sup>Created on 2022-07-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Reply by slove1200** [reply]
Modifying the time column works. Thanks @FelicienLL and @kylebaron for the reply!

**Reply by kylebaron** [reply]
Thanks @slove1200 ; best wishes for your project!
